core/src/transport/global_only.rs

```diff
@@ -0,0 +1,372 @@
+#[derive(Debug, Clone, Default)]
+pub struct GlobalIpOnly<T> {
+    inner: T,
+}
+
+mod ipv4_global {
+    use std::net::Ipv4Addr;
+
+    /// Returns [`true`] if this address part of the `0.0.0.0/8` range.
+    const fn is_this_network(a: Ipv4Addr) -> bool {
+        a.octets()[0] == 0
+    }
+
+    /// Returns [`true`] if this address part of the `0.0.0.0/32` range.
+    const fn is_this_host_on_this_network(a: Ipv4Addr) -> bool {
+        a.octets()[0] == 0 && a.octets()[1] == 0 && a.octets()[2] == 0 && a.octets()[3] == 0
+    }
+
+    /// Returns [`true`] if this address is part of the Shared Address Space defined in
+    /// [IETF RFC 6598] (`100.64.0.0/10`).
+    ///
+    /// [IETF RFC 6598]: https://tools.ietf.org/html/rfc6598
+    const fn is_shared(a: Ipv4Addr) -> bool {
+        a.octets()[0] == 100 && (a.octets()[1] & 0b1100_0000 == 0b0100_0000)
+    }
+
+    /// Returns [`true`] if this address is reserved by IANA for future use. [IETF RFC 1112]
+    /// defines the block of reserved addresses as `240.0.0.0/4`. This range normally includes the
+    /// broadcast address `255.255.255.255`, but this implementation explicitly excludes it, since
+    /// it is obviously not reserved for future use.
+    ///
+    /// [IETF RFC 1112]: https://tools.ietf.org/html/rfc1112
+    const fn is_reserved(a: Ipv4Addr) -> bool {
+        a.octets()[0] & 240 == 240 && !a.is_broadcast()
+    }
+
+    /// Returns [`true`] if this address part of the `198.18.0.0/15` range, which is reserved for
+    /// network devices benchmarking. This range is defined in [IETF RFC 2544] as `192.18.0.0`
+    /// through `198.19.255.255` but [errata 423] corrects it to `198.18.0.0/15`.
+    ///
+    /// [IETF RFC 2544]: https://tools.ietf.org/html/rfc2544
+    /// [errata 423]: https://www.rfc-editor.org/errata/eid423
+    const fn is_benchmarking(a: Ipv4Addr) -> bool {
+        a.octets()[0] == 198 && (a.octets()[1] & 0xfe) == 18
+    }
+
+    /// Returns [`true`] if this address part of the `192.0.0.0/29` range.
+    const fn is_ipv4_service_continuity_prefix(a: Ipv4Addr) -> bool {
+        a.octets()[0] == 192
+            && a.octets()[1] == 0
+            && a.octets()[2] == 0
+            && (a.octets()[3] & 0b1111_1000 == 0)
+    }
+
+    /// Returns [`true`] if this address part of the `192.0.0.0/24` range.
+    const fn is_ietf_protocol_assignments(a: Ipv4Addr) -> bool {
+        a.octets()[0] == 192 && a.octets()[1] == 0 && a.octets()[2] == 0
+    }
+
+    /// Returns [`true`] if this address part of the `192.0.0.8/32` range.
+    const fn is_ipv4_dummy_address(a: Ipv4Addr) -> bool {
+        a.octets()[0] == 192 && a.octets()[1] == 0 && a.octets()[2] == 0 && a.octets()[3] == 8
+    }
+
+    /// Returns [`true`] if this address part of the `192.0.0.9/32` range.
+    const fn is_port_control_protocol_anycast(a: Ipv4Addr) -> bool {
+        a.octets()[0] == 192 && a.octets()[1] == 0 && a.octets()[2] == 0 && a.octets()[3] == 9
+    }
```

Comment:
The entry for this one in the table you linked says "publicly reachable: true". Should we really include it in our `is_global` function then?

Reply:
@thomaseizinger I wondered the same, there are several examples like this, described as Globally Reachable, in the table:
- _Globally Reachable - A boolean value indicating whether an IP datagram whose destination address is drawn from the allocated special-purpose address block is forwardable beyond a specified administrative domain._

This explanation made me doubt then I saw a small additional notification at the top of the document which I provided in link saying: 

_Address prefixes listed in the Special-Purpose Address Registry are not guaranteed routability in any particular local or global context._

So I finally decided to put all addresses in the array as exceptions that don't meet the `is_global` condition, what do you think?

Reply:
That is a good observation. Happy to include it in the check.

I guess that should be reported to the std-lib team too? I don't see a reason why only we should extend this check.

Reply:
@thomaseizinger I've put a message [here](https://github.com/rust-lang/rust/issues/27709#issuecomment-1526091769) on the official `std` issue to have clarifications on this topic and make sure I'm not missing something. Also to be related to @mxinden's [comment](https://github.com/libp2p/rust-libp2p/pull/3814#discussion_r1178659594).

Reply:
Awesome, thank you!